src/utils.rs

```diff
@@ -92,9 +48,73 @@ pub fn to_field_element(num: BigInt, prime: BigInt) -> BigInt {
     ((num + half_prime.clone()) % prime) - half_prime
 }
 
+#[cfg(test)]
+#[macro_use]
+pub mod test_utils {
+    macro_rules! memory {
+        ( $( (($si:expr, $off:expr), $val:tt) ),* ) => {
+        {
+            let mut memory = Memory::new();
+            memory.data.push(Vec::new());
+            memory_from_memory!(memory, ( $( (($si, $off), $val) ),* ));
+        memory
+        }
+        };
+    }
+    pub(crate) use memory;
+
+    macro_rules! memory_from_memory {
+        ($mem: expr, ( $( (($si:expr, $off:expr), $val:tt) ),* )) => {
+            {
+                $mem.data.push(Vec::new());
+                $(
+                    memory_inner!($mem, ($si, $off), $val);
+                )*
+            }
+        };
+    }
+    pub(crate) use memory_from_memory;
+
+    macro_rules! memory_inner {
+        ($mem:expr, ($si:expr, $off:expr), ($sival:expr, $offval: expr)) => {
+            let mut res = $mem.insert(
+                &mayberelocatable!($si, $off),
+                &mayberelocatable!($sival, $offval),
+            );
+            while matches!(res, Err(MemoryError::UnallocatedSegment(_, _))) {
+                $mem.data.push(Vec::new());
+                res = $mem.insert(
+                    &mayberelocatable!($si, $off),
+                    &mayberelocatable!($sival, $offval),
+                );
+            }
+        };
+        ($mem:expr, ($si:expr, $off:expr), $val:expr) => {
+            let mut res = $mem.insert(&mayberelocatable!($si, $off), &mayberelocatable!($val));
+            while matches!(res, Err(MemoryError::UnallocatedSegment(_, _))) {
+                $mem.data.push(Vec::new());
+                res = $mem.insert(&mayberelocatable!($si, $off), &mayberelocatable!($val));
+            }
+        };
+    }
```

Comment:
We can make this code more brief by creating vars:
```suggestion
            let (k, v) = (&mayberelocatable!($si, $off), &mayberelocatable!($sival, $offval));
            let mut res = $mem.insert(k, v);
            while matches!(res, Err(MemoryError::UnallocatedSegment(_, _))) {
                $mem.data.push(Vec::new());
                res = $mem.insert(k, v);
            }
        };
        ($mem:expr, ($si:expr, $off:expr), $val:expr) => {
            let (k, v) = (&mayberelocatable!($si, $off), &mayberelocatable!($val));
            let mut res = $mem.insert(k, v);
            while matches!(res, Err(MemoryError::UnallocatedSegment(_, _))) {
                $mem.data.push(Vec::new());
                res = $mem.insert(k, v);
            }
        };
    }
```